llvm/lib/MC/GOFFObjectWriter.cpp

```diff
@@ -217,34 +222,403 @@ void GOFFOstream::write_impl(const char *Ptr, size_t Size) {
+}
+
+GOFFSymbol GOFFObjectWriter::createWSASymbol(uint32_t ParentEsdId,
+                                             bool ReserveQwords) {
+  const char *WSAClassName = "C_WSA64";
+  GOFFSymbol WSA = createEDSymbol(WSAClassName, ParentEsdId);
+
+  WSA.Executable = GOFF::ESD_EXE_DATA;
+  WSA.TextStyle = GOFF::ESD_TS_ByteOriented;
+  WSA.BindAlgorithm = GOFF::ESD_BA_Merge;
+  WSA.Alignment = GOFF::ESD_ALIGN_Quadword;
+  WSA.LoadBehavior = GOFF::ESD_LB_Deferred;
+  WSA.NameSpace = GOFF::ESD_NS_Parts;
+  WSA.SectionLength = 0;
+  WSA.ReservedQwords = ReserveQwords ? GOFF::ESD_RQ_1 : GOFF::ESD_RQ_0;
+
+  return WSA;
+}
+
+static uint32_t getADASectionLength(MCAssembler &Asm,
+                                    const MCAsmLayout &Layout) {
+  for (MCSection &S : Asm)
+    if (S.getName().equals(".ada"))
+      return Layout.getSectionAddressSize(&S);
+  return 0;
+}
+
+void GOFFObjectWriter::defineRootSD(MCAssembler &Asm,
+                                    const MCAsmLayout &Layout) {
+  StringRef FileName = "";
+  if (!Asm.getFileNames().empty())
+    FileName = sys::path::stem((*(Asm.getFileNames().begin())).first);
+  std::pair<std::string, std::string> CsectNames = Asm.getCsectNames();
+  if (CsectNames.first.empty()) {
+    RootSD = createSDSymbol(FileName.str().append("#C"));
+    RootSD.BindingScope = GOFF::ESD_BSC_Section;
+  } else {
+    RootSD = createSDSymbol(CsectNames.first);
+  }
+  RootSD.Executable = GOFF::ESD_EXE_CODE;
+}
+
+void GOFFObjectWriter::writeSymbolDefinedInModule(const MCSymbolGOFF &Symbol,
+                                                  MCAssembler &Asm,
+                                                  const MCAsmLayout &Layout) {
+  MCSection &Section = Symbol.getSection();
+  SectionKind Kind = Section.getKind();
+  auto &Sec = cast<MCSectionGOFF>(Section);
+
+  StringRef SymbolName = Symbol.getName();
+  // If it's a text section, then create a label for it.
+  if (Kind.isText()) {
+    auto &GSection = SectionMap.at(&Sec);
+    GOFFSymbol LD = createLDSymbol(SymbolName, GSection.PEsdId);
+    LD.BindingStrength = Symbol.isWeak()
+                             ? GOFF::ESDBindingStrength::ESD_BST_Weak
+                             : GOFF::ESDBindingStrength::ESD_BST_Strong;
+
+    // If we don't know if it is code or data, assume it is code.
+    LD.Executable = Symbol.getExecutable();
+    if (LD.isExecUnspecified())
+      LD.Executable = GOFF::ESD_EXE_CODE;
+
+    // Determine the binding scope. Please note that the combination
+    // !isExternal && isExported makes no sense.
+    LD.BindingScope =
+        Symbol.isExternal()
+            ? (Symbol.isExported() ? GOFF::ESD_BSC_ImportExport
+                                   : (LD.isExecutable() ? GOFF::ESD_BSC_Library
+                                                        : GOFF::ESD_BSC_Module))
+            : GOFF::ESD_BSC_Section;
+
+    LD.ADAEsdId = GOFFObjectWriter::ADAEsdId;
+
+    if (Symbol.getName().equals("main")) {
```

Comment:
This seems weird.  The name "main" is only special in some programming languages - if anything special is needed here, that should use a more generic mechanism.

Reply:
I'll just remove that for now, since there is still more work that has to be done to support program entry points after this.

As an aside though - it doesn't seem like there's a good mechanism for identifying the program entry point without looking for the name "main"? There are a few other places that resort to doing this, eg [here ](https://github.com/llvm/llvm-project/blob/e45f6e569dafd4033f86d276065d77799b5f6226/llvm/lib/Target/X86/X86ISelDAGToDAG.cpp#L1712 )and [here](https://github.com/llvm/llvm-project/blob/e45f6e569dafd4033f86d276065d77799b5f6226/llvm/lib/Target/PowerPC/PPCPreEmitPeephole.cpp#L415).

I do see how it's problematic to use the name "main" though but maybe something clang-side might be needed?

Reply:
Thinking about this a bit more, hard-coding a "main" name check can be justified for two reasons:
- because the language specification makes specific statements about that name (e.g. C, C++) - in this case the check should be in the language frontend (clang); or
- because the operating system / environment treats the name specially, e.g. because this is the name the process startup code (whereever this resides, startup files / C library / language environment) dispatches to - in this case the check should be in the codegen target backend (lib/Target/SystemZ) under the appropriate OS check

I still think this check shouldn't be in the GOFF writer, unless the GOFF file format as such treats a function with this name specially.

As an aside, we need to make sure the check is specific enough (e.g. there can be a static "main" function in some object which is *not* the actual "main" function with the "magic" property).

Reply:
Agreed, I think we'll move the check for `main` to the SystemZAsmPrinter. We need to know the programming language to fill in one of the PPA2 fields there anyways, so information regarding the source language should already be available there.